File: Spark-Learning/src/java/com/atguigu/spark/core/transformatioins/Cogroup.java
package com.atguigu.spark.core.transformatioins;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;

public class Cogroup {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setMaster("local");
        conf.setAppName("test");

        JavaSparkContext sc = new JavaSparkContext(conf);

        JavaPairRDD<String, Integer> nameRDD = sc.parallelizePairs(Arrays.asList(
                new Tuple2<String, Integer>("zhangsan", 18),
                new Tuple2<String, Integer>("zhangsan", 180),
                new Tuple2<String, Integer>("lisi", 19),
                new Tuple2<String, Integer>("wangwu", 20),
                new Tuple2<String, Integer>("wangwu", 200),
                new Tuple2<String, Integer>("maliu", 21)
        ));
        JavaPairRDD<String, Integer> scoreRDD = sc.parallelizePairs(Arrays.asList(
                new Tuple2<String, Integer>("zhangsan", 100),
                new Tuple2<String, Integer>("zhangsan", 1000),
                new Tuple2<String, Integer>("lisi", 200),
                new Tuple2<String, Integer>("lisi", 2000),
                new Tuple2<String, Integer>("wangwu", 300),
                new Tuple2<String, Integer>("tianqi", 400)
        ));

        JavaPairRDD<String, Tuple2<Iterable<Integer>, Iterable<Integer>>> cogroup = nameRDD.cogroup(scoreRDD);
        cogroup.foreach(new VoidFunction<Tuple2<String, Tuple2<Iterable<Integer>, Iterable<Integer>>>>() {
            @Override
            public void call(Tuple2<String, Tuple2<Iterable<Integer>, Iterable<Integer>>> stringTuple2Tuple2) throws Exception {
                System.out.println(stringTuple2Tuple2);
            }
        });

    }
}

File: Spark-Learning/src/scala/com/atguigu/sparkcore/transformation/Cogroup.scala
package com.atguigu.sparkcore.transformation

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
  * cogroup
  * 当调用类型（K,V）和（K，W）的数据上时，返回一个数据集（K，（Iterable<V>,Iterable<W>）），子RDD的分区与父RDD多的一致
  */
object Cogroup {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
    conf.setMaster("local")
    conf.setAppName("test")
    val sc = new SparkContext(conf)
    val nameRDD: RDD[(String, Int)] = sc.parallelize(Array[(String, Int)](
      ("zhangsan", 18), ("lisi", 19), ("wangwu", 20), ("maliu", 21), ("zhangsan", 180), ("maliu", 210)
    ))
    val scoreRDD: RDD[(String, Int)] = sc.parallelize(Array[(String, Int)](
      ("zhangsan", 100), ("lisi", 200), ("wangwu", 300), ("tianqi", 400),
      ("zhangsan", 1000), ("wangwu", 3000)
    ))


    val unit: RDD[(String, (Iterable[Int], Iterable[Int]))] = nameRDD.cogroup(scoreRDD)
    unit.foreach(println)
  }
}
